back/app/models/user.rb

```diff
@@ -500,15 +513,39 @@ def increment_confirmation_retry_count!
     save!
   end
 
-  def reset_email!(email)
-    update!(
-      email: email,
-      email_confirmation_code_reset_count: 0
-    )
+  def reset_email!(new_email)
+    if user_confirmation_enabled? && active?
+      update!(new_email: new_email, email_confirmation_code_reset_count: 0)
+    else
+      update!(email: new_email, email_confirmation_code_reset_count: 0)
+    end
+  end
+
+  def confirm_new_email
+    return unless new_email
+
+    self.email = new_email
+    self.new_email = nil
   end
 
   private
 
+  def validate_can_update_email
+    return unless new_email_changed? || email_changed?
+
+    if no_password? && confirmation_required?
+      # Avoid security hole where passwordless user can change when they are authenticated without confirmation
+      errors.add :email, :change_not_permitted, value: email, message: 'change not permitted - user not active'
+    elsif user_confirmation_enabled? && active? && email_changed? && !email_changed?(to: new_email_was)
```

Comment:
Also here no errors are added if the user is not active

Reply:
That is correct. If the user is not active then we only want to add errors to users with no password